rlog/2024-06-24-gsub-largemsg-improvements.mdx

```diff
@@ -0,0 +1,341 @@
+IHAVE announcements for newly received large messages trigger IWANTs from nodes 
+(including those already receiving the same message),
+leading to an additional workload for early receivers. 
+
+Potential solutions mitigating these problems include:
+1) Defering IHAVE announcements for large messages until the next heartbeat interval. 
+2) Defering IWANT requests for messages that are currently being received. 
+3) Not issuing IWANT for a message if at least $K$ peers have transmitted IDONTWANT for the same message
+(as this indicates that these peers will eventually relay this message). 
+
+#### c. IMReceiving Message
+A peer can issue an IDONTWANT only after it has received the entire message. 
+However, a large message transmission may take several hundred milliseconds to complete. 
+During this time, many other mesh members may start relaying the same message. 
+
+Therefore, the probability of simultaneously receiving the same message from multiple senders increases with the message size, 
+significantly compromising the effectiveness of IDONTWANT messages.
+
+Introducing a handshake phase for large message transmissions, 
+where the sender shares msgID and length, can provide valuable insight into ongoing message receptions. 
+
+If the receiver is already receiving this message from another sender, 
+it can request to defer this transmission. 
+
+An IDONTWANT from the receiver will indicate successful message reception. 
+Otherwise, the sender can initiate transmission after a specific wait interval. 
+
+To simplify this process, a peer can send a brief IMReceiving message (for all ongoing large message transmissions), 
+urging all mesh peers to hold off on sending the same message. 
+If no IDONTWANT is received from the peer within a specific time, 
+mesh peers can resume this transmission. 
+This approach can boost IDONTWANT benefits by also considering ongoing transmissions for large messages. 
+
+While IMReceiving messages can bring about substantial improvements in terms of latency and bandwidth utilization, 
+it's crucial to be aware of the potential risks. 
+
+This approach could be exploited by a malicious user to disrupt message transmission, 
+either by never completing a message or by intentionally sending a message at an extremely slow rate to numerous peers. 
+This could ultimately result in network-wide slow message propagation. 
+
+However, with careful calibration of the deferring interval and peer scoring, these risks can be mitigated.
+
+#### d. IDONTWANT Message with Reduced Forwarding
+It is common for slow peers to pile up outgoing message queues, 
+especially for large message transfers.
+
+This results in a significant queuing delay for outgoing messages. 
+Reduced message forwarding can help decrease the workload of slower peers. 
+
+On receiving a message longer than the specified threshold, 
+a slow peer can relay it to only $K \in D$ peers and sends an IDONTWANT message to all the peers in $D$.

```

Comment:
Is D the set of peers now rather than number of peers?

Reply:
yes, mesh peers